openmmtools/alchemy.py

```diff
@@ -1787,56 +1977,101 @@ def _alchemically_modify_NonbondedForce(self, reference_force, alchemical_region
+        for region_ids in region_permus:
+            alchemical_region_0 = alchemical_regions[region_ids[0]]
+            alchemical_region_1 = alchemical_regions[region_ids[1]]
+            if alchemical_region_0.alchemical_atoms == alchemical_region_1.alchemical_atoms:
+                if alchemical_region_0.region_name == None:
+                    #case where only single populated region which is unnamed
+                    region_names = ['']
+                else:
+                    region_names = ['_' + alchemical_region_0.region_name]
+            else:
+                region_names = ['_' + alchemical_region_0.region_name, '_' + alchemical_region_1.region_name]
+            # --------------------------------------------------
+            # Determine energy expression for all custom forces
+            # --------------------------------------------------
 
-        # Define energy expression for sterics.
-        sterics_energy_expression = exceptions_sterics_energy_expression + sterics_mixing_rules
+            sterics_mixing_rules, exceptions_sterics_energy_expression = self._get_sterics_energy_expressions(region_names)
 
-        # Define energy expression for electrostatics based on nonbonded method.
-        nonbonded_method = reference_force.getNonbondedMethod()
-        is_ewald_method = nonbonded_method in [openmm.NonbondedForce.Ewald,
+            # Define energy expression for sterics.
+            sterics_energy_expression = exceptions_sterics_energy_expression + sterics_mixing_rules
+
+            # Define energy expression for electrostatics based on nonbonded method.
+            nonbonded_method = reference_force.getNonbondedMethod()
+            is_ewald_method = nonbonded_method in [openmm.NonbondedForce.Ewald,
                                                openmm.NonbondedForce.PME]
-        is_rf_method = nonbonded_method in [openmm.NonbondedForce.CutoffPeriodic,
+            is_rf_method = nonbonded_method in [openmm.NonbondedForce.CutoffPeriodic,
                                             openmm.NonbondedForce.CutoffNonPeriodic]
-        is_periodic_method = is_ewald_method or nonbonded_method == openmm.NonbondedForce.CutoffPeriodic
-        use_exact_pme_treatment = is_ewald_method and self.alchemical_pme_treatment == 'exact'
+            is_periodic_method = is_ewald_method or nonbonded_method == openmm.NonbondedForce.CutoffPeriodic
+            use_exact_pme_treatment = is_ewald_method and self.alchemical_pme_treatment == 'exact'
 
-        # Warn about reaction field.
-        if is_rf_method:
-            logger.warning('Reaction field support is still experimental. For free energy '
+            # Warn about reaction field.
+            if is_rf_method:
+                logger.warning('Reaction field support is still experimental. For free energy '
```

Comment:
These flags (from here up to `# Define energy expression for electrostatics based on nonbonded method.`) depend only on the reference `NonbondedForce` and can be moved outside the loop.